### PR TITLE
RFC Code of Conduct

### PR DESCRIPTION
**Pull request by tiborsimko**
Our community is growing, and as part of improving our community documentation #1799, we thought of adopting a Code of Conduct.   This PR proposes a text for discussion.

Design guidelines:
  - should be short, sweet, no-nonsense
  - should promote good collaborative development principles

Proposal:
 - endorse literally the Python Code of Conduct
 - recommend literally the egoless programming principles (TPOCP)

For the full text, please see the file `docs/community/code-of-conduct.rst`that  comes with this
PR.

Please comment on the text, upvote, downvote, or otherwise express your opinions.

### CODE REVIEW

File: docs/community/code-of-conduct.rst [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+   3. **Respectful.** Members of the community are respectful.  We're
+   respectful of others, their positions, their skills, their
+   commitments, and their efforts.  We're respectful of the volunteer
+   efforts that permeate the Invenio community.  We're respectful of
+   the processes set forth in the community, and we work within them.
+   When we disagree, we are courteous in raising our issues.
+
+   Overall, we're good to each other.  We contribute to this community
+   not because we have to, but because we want to.  If we remember
+   that, these guidelines will come naturally.
```
Comment (kaplun) [inline docs/community/code-of-conduct.rst]: I have only one remark on this very last point: although Invenio is fully Open Source, it is true that many developers are salaried to work on a service that is based on Invenio. Therefore they are genuinely starting from a position where they contribute to this community because they have to. However the wish is that, as time goes by, they are getting more and more into wanting to be part of the community.

As my daughters' teachers explained us about their school program: "*We don't want that the children are simply doing what they want, rather that they want what they are doing*".  

Additionally I would add  a 4 point.

4. **Generous.** Members of the community are generous. They think of their peers needs in terms of information, code, features but also personal care, consideration, couching, time. The more you give, the more you will receive.

**Comment by lovasko** [comment]
> we thought of adopting a Code of Conduct.

what are the /goals of/reasons for/ such document?

**Comment by kaplun** [comment]
> what are the /goals of/reasons for/ such document?

We are currently nearly 30 developers and overall (according to GitHub) 89 historical contributors. Although many of us are based at CERN, some are elsewhere, e.g. in Germany or Greece, or US...

Sometimes it's not easy to solve disputes being in the same room, imagine in a middle-large and distributed community. 

Such document could be a sort of mantra that everyone could keep available (e.g. printed on a wall), and to be checked during hard times (but also the good ones), and that could help fostering a community spirit in this project, if it was really followed by everyone.

**Comment by lovasko** [comment]
Are there any points that diverge from a common sense or a basic human interaction rules? Also, is this initiative addressing a _specific_ problem?

**Comment by kaplun** [comment]
> Are there any points that diverge from a common sense or a basic human interaction rules? 

I don't know. Do you find easy to endorse *all* of the points? Then you are a super hero :-) E.g. I for once I have personally some difficulties with at least point 9:
*Don't be the guy coding in the dark office emerging only to buy cola.*.

> Also, is this initiative addressing a specific problem?

As far as I know, it stems simply from the fact that is come the moment to have such a thing. Invenio is maturing, we have 2.0, we are rewriting/redesigning it for the good, and trying to follow mainstreams Open source conventions, such as e.g. having an internal code of conduct for the community.

**Comment by lovasko** [comment]
> Do you find easy to endorse all of the points?

No, I fail from time to time, just as everyone else. But I don't see them being written in a GitHub repository file to help me in any way, just because I already know all of them (just because it is common sense, as mentioned).

The question is: is it just going to be a text that is floating around, or is it going to be enforced in any way? If so, isn't that just adding complexity to the development process?

> it stems simply from the fact that is come the moment to have such a thing
I still fail to see the _need_ here. If there is no particular issue being solved, we are benevolently adding more clutter to the communication process. And this is a thing that should go exactly the other way around.

**Comment by kaplun** [comment]
> just because I already know all of them (just because it is common sense, as mentioned).

I don't think these are necessarily common sense. The fact that you know all of them, shows only that you in particular are already having experience in Open source.

But e.g. the egoless coding principles go against most of the common sense, e.g. it is very likely not instinctively shared by high-school bullies, larger-than-life-egos people, etc.

People are very different in terms of experiences, background and culture...

**Comment by jirikuncar** [comment]
@lovasko @kaplun you both have interesting points in this conversation.

1. the *code* is here to be discussed;
2. its content might be a subject of change (e.g. the **Generous** point 4);
3. we can't really force people to follow it but we can ask them to acknowledge its existence by e.g. adding `Signed-off-by` signature to their commits (*just an example!*).

@tiborsimko I'm fine with your proposal, I just would *not* like to see it longer.

**Comment by glouppe** [comment]
While I have nothing against a code of conduct per se, I am not sure this is really necessary either. So +0 for adding one in general.  

If a code of conduct is adopted though, I think this one is too long and the 3 first points more or less encompass all 10 others. As another example, have a look at https://engineering.twitter.com/opensource/code-of-conduct

One thing that I would like to be clarified is whether and how Invenio is open for others to take responsibilities. I really believe that for a transition towards community-driven development  to be effective (assuming this is where Invenio wants to go?), the project needs to make its contributors feeling involved and responsible. In my opinion, an open source project is more than a project with an open source license. An open source project is a project where its contributors feel like they can become a part of it, like they can take decisions and take responsibilities. For instance, with regards to Invenio, I believe more contributors should have ownership rights of the repository, more contributors should feel like they are equal. More contributors should be let making mistakes and learn from them, even if founders think otherwise. In this way (but there may be others), I feel like contributors would stay more motivated and more contributors would keep contributing in the long term. 

In summary, such a statement towards openness is not clear for me with respect to the proposed code of conduct, nor I know if this is something everyone agrees on. I think Sam's teacher is right on: "We don't want that the children are simply doing what they want, rather that they want to do what they are doing." 

Just my 2 cents :)

**Comment by tiborsimko** [comment]
> what are the /goals of/reasons for/ such document?

In addition to good reasons already highlighted by @kaplun, there were a few cases of tension in the past, as probably in any larger group. Since we are writing up the Community documentation section, it is a good occasion to formalise an informal Code of Conduct document, reminding people to respect each other's karate.  In my experience not many of our newcomers have heard of "egoless programming" principles.  If you feel they are only transcribing "common sense", than so much the better.

File: docs/community/code-of-conduct.rst (hunk 1 above)
Comment (tiborsimko) [inline docs/community/code-of-conduct.rst]: > although Invenio is fully Open Source, it is true that many developers are salaried to primarily work on a service that happens to be based on Invenio. Therefore they are genuinely starting from a position where they contribute to this community because they have to.

Nuance: there are notably more Invenio contributors who are payed to work on various services than there are developers who are payed to work on the Invenio core itself.  Hence occasional dilemmas like "do I make this feature more general for others to directly reuse or do I stop here at my service boundaries?" which is a sort of "want to", not "have to" situation.

However, please note that the above phrase comes from the Python Code of Conduct "as is"; I have not changed anything.  I find it simpler to endorse some existing text rather than start inventing our own.

> Additionally I would add a 4th point. [...] Generous.

While I like the spirit of your point, I'm somewhat concerned that we would be modifying Python's vanilla Code of Conduct instead of citing it literally.  People may then question why this-or-this was added, removed, or modified.  A certain elegance behind my proposal, if I may say so, was that we would be simply endorsing two "well established" sets of guiding principles, which I hoped wouldn't be too controversial.  (Especially since Invenio is a Pythonic project.)  So, instead of modifying Python's upstream Code of Conduct, I'd rather prefer if you point to some other well established standard text, if you feel like adding an explicit point about generosity?

**Comment by tiborsimko** [comment]
> we can ask them to acknowledge its existence by e.g. adding Signed-off-by signature to their commits (just an example!).

It would seem like an overkill to use `Signed-off-by` with every single branch that people submit to mean that they acknowledge the general Code of Conduct.   It should be enough to get acquainted with the Code of Conduct and acknowledge it once, like during the boot camp, as it won't be changing often.  The individual branch signatures would be better reserved to something specific that can vary for each concrete branch, such as QA, though let's see in #2876.

**Comment by tiborsimko** [comment]
> One thing that I would like to be clarified is whether and how Invenio is open for others to take responsibilities.

This will come into the same Community documentation section as part of #1799 efforts.  In the meantime, the general principles have been written up elsewhere, see e.g. [triage docs](https://forum.invenio-software.org/t/triaging-issues-labels-milestones/58/6).  In a one-line summary, people can take on responsibilities by meritocracy: start contributing to the ecosystem, commenting on issues, reviewing people's branches, helping out with bug fixes and tests, not only in "your modules" but in general, and see your project responsibilities widen.  The above URL highlights principles on how we triage issues and prioritise milestones; similar will follow on how we merge, etc.

**Comment by dset0x** [comment]
@jirikuncar happened to be in my office a few days ago and kindly reminded me that I should try to be more active in maintaining my PRs. I think raises is an important point: responsibility. We all have different positions here - some with low expectancies - but being an open source project we should at least try to find a personal point in what we do and motivate ourselves around it.

During a brief search I stumbled upon [ACM's Code of Ethics](http://www.acm.org/about/code-of-ethics) and I found paragraphs 1.3 and 2.6 to relate to my idea. The document is definitely broader in range, but I think it is a good read even if we don't end up integrating it.

---

Now, not only I think we should have this code of conduct, we should also hang it on our walls to remind ourselves how we should balance community, pleasure and work.

We should also not fear to reference this code and improve upon it. Hurting someone's feelings temporarily for the good of the entire community is a useful and valuable lesson for all.

File: docs/community/code-of-conduct.rst (hunk 1 above)
Comment (kaplun) [inline docs/community/code-of-conduct.rst]: > While I like the spirit of your point, I'm somewhat concerned that we would be modifying Python's vanilla Code of Conduct instead of citing it literally.

Yep, but indeed I wanted to stress that we start from a somehow different community than the Python one and therefore we might need to acknowledge this into our code of conduct. Take it like an inheritance from Python code of conduct and where we override some of its methods :-) and make it more ours...

**Comment by lnielsen** [comment]
+1

Just some few comments on points mentioned above:

1. **Need:** We are a growing community and the code of conduct is a reminder to as all of the basic rules of interaction. Not only as person to person, but also just as important: service to service (Inspire vs Zenodo vs JUSER, TIND vs CERN, etc.). While perhaps obvious in theory, in practice I think it's far from so. I've personally experience many tensions where it would have been good to have a CoC as a reminder.
1. **Code of Conduct (CoC) vs Egoless programming**: I think it would be useful to make it into two sections, so that the CoC is very short, and then stick egoless programming under a section called something like "Good Behavioural Practices".
1. **Enforcement**: It might be useful to have a part saying that violations of the CoC can be reported to any member of the triaging team or similar.
1. **Python vs custom**: I agree with @tiborsimko that a literal adoption of an existing CoC is preferable. A bit like with using other Python packages: let's not reinvent the wheel. Lots of thought has gone into the right wording of the existing CoCs.
1. **Open Source vs Open Collaboration**: I fully agree with @glouppe and @kaplun that we need what I would call an "open collaboration" where **you want** to contribute. They keyword to an open collaboration I think is **frictionless**. Technically frictionless to get Invenio up running, and socially frictionless to contribute and take part. We have been addressing the former the past years, and the CoC is just the first steps in addressing the latter. I believe that clear playing rules is needed in an open collaboration - it's needed for newcomers to know how they can eventually become integrators and take decisions, and it's needed for long standing contributors to give up control to the community.

**Comment by lovasko** [comment]
@tiborsimko we made a LaTeX version of the document for a nice print version

**Comment by jirikuncar** [comment]
@lovasko `rst2latex` ;)

**Comment by jirikuncar** [comment]
@tiborsimko can we merge it as is? There were no major objections and more suggested documents might come in upcoming days.

--
Following documents should be proposed and discussed in future RFCs:
- [ ] "Good Behavioural Practices"
- [ ] Pull-Request Integration Checklist
- [ ] Release Checklist
- [ ] "How to become triager/integrator"@tiborsimko can we merge it as is? There were no major objections and more suggested documents might come in upcoming days.

--
Following documents should be proposed and discussed in future RFCs:
- [ ] "Good Behavioural Practices"
- [ ] Pull-Request Integration Checklist
- [ ] Release Checklist
- [ ] "How to become triager/integrator"

**Comment by tiborsimko** [comment]
> Following documents should be proposed and discussed in future RFCs

This is indeed coming as part of #1799, as I mentioned above.  (By simply centralising current practices into one place; any possible RFCs could be future amendments on top of that.)

**Comment by tiborsimko** [comment]
> "Good Behavioural Practices"

This sounds a lot like a "Code of Conduct".  I'd simply leave them together, people should have enough bandwidth to read 13 short points in one single place...

**Comment by jirikuncar** [comment]
@tiborsimko `docs/community/code-of-conduct.rst:: WARNING: document isn't included in any toctree`

**Comment by tiborsimko** [comment]
@jirikuncar Yes, the inclusion of CoC in the overall docs is in my other branch that introduces "Community" docs structure.  (That's by design: the CoC body was best discussed separately, while for the overall docs structure we have #1799 already.)  I'll issue another PR for the Community docs afterwards.

**Comment by jirikuncar** [comment]
@tiborsimko we can't merge it because the tests with fail to build docs in strict mode (https://travis-ci.org/inveniosoftware/invenio/jobs/57621067).

**Comment by tiborsimko** [comment]
OK, I run sphinx without `-W` locally..  I'll amend this.
